File: src/main/java/Day5.kt
class Day5: Day<IntArray> {

    override fun parse(text: String): IntArray {
        return text.split("\n").map { it.trim() }.filter { !it.isEmpty() }.map { it.toInt() }.toIntArray()
    }

    override fun part1(input: IntArray): Any {
        return solve(input, { _, _ -> 1 })
    }

    fun solve(input: IntArray, valueModification: (IntArray, Int) -> Int): Int {
        var idx = 0
        var count = 0
        while (true) {
            val oldIdx = idx
            val value = input[oldIdx]
            count++
            idx += value
            if (idx < 0 || idx >= input.size) {
                break
            }
            input[oldIdx] += valueModification.invoke(input, value)
        }
        return count
    }

    override fun part2(input: IntArray): Any {
        return solve(input, { _, v -> if (v >= 3) -1 else 1 })
    }


}

File: 2019-typescript/src/Day5.ts
import Day from "./Day";
import TestCase from "./TestCase";
import IntCodeComputer from "./IntCodeComputer"

class Day5 implements Day<number[]> {

  parse(text: string): number[] {
    return text.split(',').map((x: string) => parseInt(x, 10));
  }

  part1(input: number[]): any {
    let computer = new IntCodeComputer()
    computer.runProgram(input);
    return computer.outputs[computer.outputs.length - 1];
  }

  part2(input: number[]): any {
    return 0;
  }

  testCases: TestCase[] = [
      {
        text: "1002,4,3,4,33",
        expectInput: [ 1002, 4, 3, 4, 99 ],
        result: ""
      },
      {
        text: "1001,1,-106,1,99",
        expectInput: [ 1001, -105, -106, 1, 99 ],
        result: ""
      }
    ]
}

export default Day5;
